### PR TITLE
PagerShapes: Use IShapeFactory and cleanup

### PR DESCRIPTION
**Pull request by jtkech**
Maybe related to #8887

In my case to repro

- Setup with the blog recipe
- Create more than 10 blog posts
- Go to the blog home page that list the first 10 blog posts
- Click on Older posts => you get the following exception

Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware|ERROR|An unhandled exception has occurred while executing the request. System.**ArgumentOutOfRangeException**: Specified argument was out of the range of valid values. (Parameter 'index')
   at Microsoft.AspNetCore.Mvc.ViewFeatures.Buffers.ViewBufferTextWriter.Write(Char[] buffer, Int32 index, Int32 count)
   at System.IO.TextWriter.Write(ReadOnlySpan`1 buffer)
   at System.IO.TextWriterExtensions.WritePartialString(TextWriter writer, String value, Int32 offset, Int32 count)
   at System.Text.Encodings.Web.TextEncoder.Encode(TextWriter output, String value, Int32 startIndex, Int32 characterCount)
   at System.Text.Encodings.Web.TextEncoder.Encode(TextWriter output, String value)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.**StringHtmlContent.WriteTo**(TextWriter writer, HtmlEncoder encoder)
   at Microsoft.AspNetCore.Html.**HtmlContentBuilder.WriteTo**(TextWriter writer, HtmlEncoder encoder)
   at Microsoft.AspNetCore.Mvc.Rendering.**TagBuilder.WriteTo**(TagBuilder tagBuilder, TextWriter writer, HtmlEncoder encoder, TagRenderMode tagRenderMode)

So in `PagerShapes` it fails when adding an `StringHtmlContent` in the inner `HtmlContentBuilder` of a `TagBuilder`

It fails under liquid, i also tried it with the dev branch before Fluid 2.0, it also fails so not related.

Replacing the `StringHtmlContent` by a copy here named `HtmlContentString` fixed the issue

The only difference is that the original `StringHtmlContent` does in its `WriteTo()`

    encoder.Encode(writer, _input);

With our new `HtmlContentString` here we do

    writer.Write(encoder.Encode(_value))

### CODE REVIEW

**Comment by jtkech** [comment]
@sebastienros 

In fluid the same change has been done for the `StringValue`on Feb 6, 2021
https://github.com/sebastienros/fluid/blob/ad942870db05bf7f57b23e4d77fa628ec780a891/Fluid/Values/StringValue.cs#L149-L156

                // perf: Don't use this overload
                // encoder.Encode(writer, _value);

                // Use a transient string instead of calling
                // encoder.Encode(TextWriter) since it would
                // call writer.Write on each char if the string
                // has even a single char to encode
                writer.Write(encoder.Encode(_value));

So maybe worth to replace all our `StringHtmlContent` usages with a new custom `HtmlContentString` as described in my previous comment, doing a `writer.Write(encoder.Encode(_value))` in place of a `encoder.Encode(writer, _value)`.

At least it fixed the pager issue under liquid

Then yes, would be better to fully understand why it was failing

I will look in the source code of writers / encoders, what does `.WritePartialString()` as seen in the above trace.

**Comment by jtkech** [comment]
So just to confirm that it is related to encoding

If for the pager i don't use any special char, e.g. `Older Posts` in place of `Older Posts →`, it doesn't fail

Can't find the source code of `System.IO.TextWriterExtensions.WritePartialString()`, looks like a new thing

So i think i will merge this one for the cleanup part of the pager, as it uses `IShapeFactory` in place of the dynamic `New`, and as it already fixes an issue when used under liquid

Then i will open a separate PR suggesting to replace all the `StringHtmlContent` by a new custom `HtmlContentString` using `writer.Write(encoder.Encode(_value))` in place of a `encoder.Encode(writer, _value)` as anyway it is beter to do that.

**Comment by sebastienros** [comment]
System.Text.Encodings.Web was updated with the latest runtime release, for a security fix. There might be a bug in it, but I would bet for a bug in our own buffers. The startIndex out of range exception and the fact that we see buffer overflows in other places is the same bug: the offset is bigger than the expected string, and either the array is too small and there is an exception, or the array is big enough and it show some stale content from a previously pooled buffer.

So probably that some pooled buffer is not correctly initialized, or size is miscalculated. The only strange thing is that it's triggered by the Encoder update. I will check the changes.

File: src/OrchardCore.Modules/OrchardCore.Navigation/PagerShapes.cs [hunk 1]
```diff
@@ -550,7 +572,16 @@ private IHtmlContent CoerceHtmlString(object value)
                 return result;
             }
 
-            return new StringHtmlContent(value.ToString());
+            return new HtmlContentString(value.ToString());
+        }
+
+        private class HtmlContentString : IHtmlContent
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Navigation/PagerShapes.cs:578]: Why a new class since we already have it?

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Navigation/PagerShapes.cs:578]: Or is it that in the meantime you created it ;)

**Comment by madelson** [comment]
@sebastienros we aren't using Orchard but are seeing the same error after updating the encodings library. I'm curious, did you ever trace this down or report to MSFT?

**Comment by sebastienros** [comment]
@madelson yes, it's well understood and reported (PS: I work for MSFT). Can you comment on https://github.com/dotnet/aspnetcore/issues/31299 and explain how you use it to repro the issue? Because OC was the only app reporting the issue, and there was a mitigation ready, this didn't get fix in a 5.0 service release.

**Comment by sebastienros** [comment]
@madelson though if your issue is not related to an "empty" call to Encode(), then you probably have an issue that the Encoder now rents the buffers, so you should not keep a reference to them. This is a change in behavior from the .NET side.

**Pull request merged**